### ISSUE TITLE
Use `aria-modal` in dialogs

### ISSUE THREAD

**Issue by tom-sherman** [mui/base-ui#4843]
# Feature request

## Summary

`aria-modal` is a more modern alternative

## Examples in other libraries

React Spectrum uses aria-modal (with an optional [polyfill](https://github.com/adobe/react-spectrum/tree/092229c62c070e6405d4b6d74bb27f328508e781/packages/%40react-aria/aria-modal-polyfill))

## Motivation

Using `aria-modal` should present a better UX for users using browser/screen readers that support this property compared to adding `aria-hidden` to elements outside the dialog.

**Comment by atomiks** [comment]
The current `aria-hidden` approach has an important exception for live regions: `[aria-live]` elements aren't hidden, so announcers like `<Toast.Viewport>` can continue to announce while a dialog is open.

If we rely only on `aria-modal="true"`, AT can treat live announcers outside the dialog as hidden. That means toasts and other announcers fail to announce while the dialog is open - there’s discussion in w3c/aria#1854, and the behavior does not seem consistent across ATs.

> React Spectrum uses aria-modal

As far as I can see, they actually don't use it, though only due to a "Safari bug":

> // **We do not use aria-modal** due to a Safari bug which forces the first focusable element to be focused
// on mount when inside an iframe, no matter which element we programmatically focus.
// See https://bugs.webkit.org/show_bug.cgi?id=211934.
// useModal sets aria-hidden on all elements outside the dialog, so the dialog will behave as a modal
// even without aria-modal on the dialog itself.

Also React Aria preserves them only through its own conventions: `data-live-announcer` and `data-react-aria-top-layer`. It does not seem to preserve arbitrary `[aria-live]` regions outside the modal unlike ours. Base UI’s current implementation is more permissive because it keeps generic `[aria-live]` nodes visible to AT
